File: sensors/src/min_2Dscan.cpp
#include <ros/ros.h>
#include <pcl_ros/point_cloud.h>
#include <pcl/point_types.h>
#include <pcl/conversions.h>
#include <boost/foreach.hpp>
#include "sensor_msgs/LaserScan.h"
#include <math.h>

typedef pcl::PointCloud<pcl::PointXYZ> PointCloud;
ros::Publisher pub_scan; // Public or private?

void callback(const PointCloud::ConstPtr& msg)
{
    // Init variables
    float x;
    float y;
    float z;
    float depth;
    float angle;
    int idx;
    sensor_msgs::LaserScan lscan;

    // Init the laserscan msg
    ros::Time stmp = pcl_conversions::fromPCL(msg->header).stamp;
    lscan.header.stamp = stmp;
    lscan.header.frame_id = "camera";
    lscan.angle_min = -M_PI/4;
    lscan.angle_max = M_PI/4;
    lscan.angle_increment = M_PI*0.001;
    lscan.range_min = 0.3;
    lscan.range_max = 7.0;
    std::vector<float> init_ranges(500,7.0);
    lscan.ranges = init_ranges;

    // Process the ranges by iterating through each point of the pointcloud
    BOOST_FOREACH (const pcl::PointXYZ& pt, msg->points) {
        x=-pt.x;
        y=-pt.y;
        z=pt.z; // Get value directly, not the pointer?
        if (!(isnan(x) || isnan(y) ||isnan(z)) && z>0.05 && y<0.2 && y>-0.2){
            // Then the point is within the robot's height and is in front of it.

            // Process depth and angle
            depth = std::sqrt(std::pow(x,2) + std::pow(z,2));
            if (depth>lscan.range_min && depth<lscan.range_max){ // Discard depths outside [min,max]
                angle = std::atan(x/z); // atan2 not needed as the point is in z>0
                idx = std::floor((angle - lscan.angle_min) / lscan.angle_increment);
                if (lscan.ranges[idx]>depth) {
                    lscan.ranges[idx]=depth;
                }
            }
        }
    }
    pub_scan.publish(lscan);

}

int main(int argc, char** argv)
{
  ros::init(argc, argv, "min_2Dscan");
  ros::NodeHandle nh;
  ros::Subscriber sub = nh.subscribe<PointCloud>("pcl", 1, callback);
  pub_scan = nh.advertise<sensor_msgs::LaserScan>("min_scan/scan", 1);
  ros::spin();
}

File: sensors/scripts/min2Dscan.py
#!/usr/bin/env python3

import rospy
from sensor_msgs.msg import PointCloud2
import sensor_msgs.point_cloud2 as pc2
from sensor_msgs.msg import LaserScan

from numpy import pi
import numpy as np
import ros_numpy

"""
    This script provides a lidar scan message based on the /pointcloud
    received information.
    The algorithm considers ona angle at a time, and choses the closest depth
    among the points at that angle, between min_height and max_height (respective to the camera's height).
"""


class MinScan:
    def __init__(self):
        rospy.init_node("min_scan", anonymous=True)
        self.scan_publisher = rospy.Publisher("min_scan/scan", LaserScan, queue_size=1)
        self.cloud_subscriber = rospy.Subscriber("/pointcloud", PointCloud2, self.cloudCallback)
        self.pc_pub = rospy.Publisher("min_scan/pc", PointCloud2, queue_size=1)

    def filter_point_cloud(self, np_pc):
        # In this frame, x is the width, y is the height and z is the depth
        min_height = -0.3  # Respect to the camera's height
        max_height = 0.4

        pc_x = -np_pc["x"]
        pc_y = -np_pc["y"]
        pc_z = np_pc["z"]

        nan_filter = ~np.isnan(pc_x) & ~np.isnan(np_pc["y"]) & ~np.isnan(pc_z)
        pc_x, pc_y, pc_z = pc_x[nan_filter], pc_y[nan_filter], pc_z[nan_filter]

        x_filter = np.abs(pc_x) < (6.0 / 2)
        y_filter = (pc_y < max_height) & (pc_y > min_height)
        z_filter = pc_z > 0.05
        pc_filter = x_filter & y_filter & z_filter

        pc_x, pc_y, pc_z = pc_x[pc_filter], -pc_y[pc_filter], pc_z[pc_filter]
        return pc_x, pc_y, pc_z

    def cloudCallback(self, msg):

        start_time = rospy.Time.now().to_nsec()
        # Variables to make params from
        a_increment = pi * 0.001

        # Read the point cloud data
        np_pc = ros_numpy.numpify(msg)
        pc_x, pc_y, pc_z = self.filter_point_cloud(np_pc)

        # Filter the 2D line out of the 3D point cloud
        # points = []
        # for k in range(len(pc_x)):
        #     points.append((pc_x[k], pc_y[k], pc_z[k], 1))
        # head = msg.header
        # head.frame_id = "camera_left"
        # pc_msg = pc2.create_cloud(head, msg.fields, points)

        # Publish the "2D point cloud"
        # self.pc_pub.publish(pc_msg)

        # Process angles and depths
        angles = np.arctan(np.multiply(pc_x, 1 / pc_z))
        depths = np.sqrt(pc_x**2 + pc_z**2)

        # Init the LaserScan msg
        lscan = LaserScan()
        lscan.header.stamp = msg.header.stamp
        lscan.header.frame_id = "camera"
        lscan.angle_min = np.min(angles)
        lscan.angle_max = np.max(angles)
        lscan.angle_increment = a_increment
        lscan.range_min = np.min(depths)
        lscan.range_max = np.max(depths)
        lscan.ranges = [0.0] * int((lscan.angle_max - lscan.angle_min) / a_increment + 1)

        # Actually compute the ranges
        for idx in range(len(lscan.ranges)):
            # For each laserscan direction, keep the minimal depth among the points closer to that direction
            points_set = np.floor((angles - lscan.angle_min) / a_increment).astype(int) == idx
            if points_set.any():
                lscan.ranges[idx] = np.min(depths[points_set])
            else:
                lscan.ranges[idx] = lscan.range_max

        # Publish it
        self.scan_publisher.publish(lscan)
        rospy.loginfo("Processed in: " + str(rospy.Time.now().to_nsec() - start_time))

    def listen(self):
        rospy.spin()


min_scan = MinScan()
min_scan.listen()
